pilot/pkg/serviceregistry/kube/controller/endpoints.go

```diff
@@ -0,0 +1,224 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pkg/config/schemas"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/serviceregistry/kube"
+	configKube "istio.io/istio/pkg/config/kube"
+	"istio.io/istio/pkg/config/labels"
+	"istio.io/pkg/log"
+)
+
+// Pilot can get EDS information from Kubernetes from two mutually exclusive sources, Endpoints and
+// EndpointSlices. The edsController abstracts these details and provides a common interface that
+// both sources implement
+type edsController interface {
```

Comment:
may be move this to `endpointslice.go` as that seems to be the longer term thing, and eventually this gets deprecated/deleted?

Also there seems to some code overlap between these controllers - do you think it is worth refactor and bring some common functions out or since this will eventually be deleted, is it OK to leave the code in two places?

Reply:
Maybe even just move this out to its own file.  Maybe we could do something like:

endpoint_discovery.go // interface
endpoints_controller.go
endpointslice_controller.go

In general I prefer spelling it out rather than using the abbreviation `eds`.

Reply:
the reason  I went with eds is to avoid confusing with endpoints (the concept) with Endpoints the API. calling it eds maybe doesn't help much there though

Reply:
>  eventually this gets deprecated/deleted

I don't think there is any plan to ever remove Endpoints given they are a `v1` API

Reply:
I broke this out as suggested

Reply:
> I don't think there is any plan to ever remove Endpoints given they are a v1 API

They will not be removed from K8s API, But, in Istio do we want to support both in the longer term?